File: work/ii_estimate_SRO_and_pACS.m
% Estimate SRO with DXCP PhaT 
% Estimate prelim. ACS
%
%


%% INIT

clear variables; %close all; %clc;
conf = Config();


%% PARAMETERS: GENERAL

dbPath = conf.pathDB_test;%conf.pathDB_train;
load([dbPath, 'parameters.mat'], 'fs_Hz', 'sigLen_s', 'n_setups');
sigLen_smp = sigLen_s*fs_Hz;


%% PARAMETERS: DXCP-PhaT

parDXCPPhaT = sroest.DXCP_PhaT_DefaultParams();
sigLen_frames = floor((sigLen_smp - parDXCPPhaT.FFTsize)/parDXCPPhaT.FFTshift) + 1;


%% ESTIMATE SRO AND PRE-ACS

% Prepare Output (save results matrix-style)
SRO_Est = zeros(n_setups, sigLen_frames); 
chi = zeros(n_setups, sigLen_frames); 

% Loop Iteration parameters and load corresponding signals
parfor nn = 1:n_setups
    %Load signals
    filename = [dbPath, 'mat_files/setup_', num2str(nn), '.mat'];
    data = load(filename, 'sig_z');
    sig_z = data.sig_z;
    tic;
    %Estimate SRO (open-loop)
    disp(['Processing for setup ' num2str(nn)]);
    [SROest_ppm, ~, ~, ~, ~, ~, ~, ~, GCSD2_avg_hist, ~] = sroest.DXCP_PhaT_Jrnl_mod(sig_z, parDXCPPhaT);       
    SRO_Est(nn, :) = SROest_ppm;
    % Compute chi a.k.a. prelim. ACS
    [~, ~, preACS, ~] = acs.acs(GCSD2_avg_hist(1:4097,:), "EXP", "bin");
    chi(nn, :) = preACS;
    toc;     
end    


%% SAVE RESULTS

if ~isfolder([dbPath, 'results'])
    mkdir([dbPath, 'results']);
end
if ~isfile([dbPath '/results/est_results.mat'])
    save([dbPath '/results/est_results.mat'], 'SRO_Est', 'chi');
else
    error('Warning: Did not save results. A file with the same name already exists!');
end

File: work/python/ii_estimate_SRO_and_pACS.py
#%% 
import numpy as np
import os
import scipy.io as sio
from scipy import signal
from modules.sro_estimation import DXCPPhaT
from modules.acs import ACS
from tqdm import tqdm
import matplotlib.pyplot as plt
from config import *

# %% PARAMETERS: GENERAL

db_path = PATH_DB_TRAIN  # or conf.pathDB_train
data = sio.loadmat(os.path.join(db_path, 'parameters.mat'), mat_dtype=True)
fs_Hz = data['fs_Hz'][0][0]
sigLen_s = data['sigLen_s'][0][0]
n_setups = int(data['n_setups'][0][0])
sigLen_smp = sigLen_s * fs_Hz

# %% PARAMETERS: DXCP-PhaT
parDXCPPhaT = DXCPPhaT.defaultParams
sigLen_frames = int(np.floor((sigLen_smp - parDXCPPhaT["FFTsize_dxcp"]) / parDXCPPhaT["FFTshift_dxcp"]) + 1)


# %% ESTIMATE SRO AND PRE-ACS

# Prepare Output (save results matrix-style)
SRO_Est = np.zeros((n_setups, sigLen_frames))
chi = np.zeros((n_setups, sigLen_frames))

# Loop Iteration parameters and load corresponding signals
for nn in range(n_setups):

    dxcpPhaT = DXCPPhaT()
    acs = ACS()

    # Load signals
    filename = os.path.join(db_path, f'mat_files/setup_{nn+1}.mat')
    data = sio.loadmat(filename)
    sig_z = data['sig_z']
    print(f'Processing for setup {nn+1}')

    sig_z_frames = np.lib.stride_tricks.sliding_window_view(sig_z, window_shape=(parDXCPPhaT["FFTshift_dxcp"],2))[::parDXCPPhaT["FFTshift_dxcp"]]

    for ii in tqdm(range(sig_z_frames.shape[0])):
    #for ii in tqdm(range(100)):

        res = dxcpPhaT.process_data(sig_z_frames[ii,0,:,:])
        if ii <= 2: # first 2^13 FFT analysis frame is available only after three iterations
            continue
        acs.process(res['GCSD2_avg'][:4097])

        SRO_Est[nn, ii-3] = res['SROppm_est_out']
        chi[nn, ii-3] = acs.preACS



# %% SAVE RESULTS

if True:
    results_path = os.path.join(db_path, 'results_python')
    if not os.path.isdir(results_path):
        os.makedirs(results_path)
    results_file = os.path.join(results_path, 'est_results_python.mat')

    if not os.path.isfile(results_file):
        sio.savemat(results_file, {'SRO_Est': SRO_Est, 'chi': chi})
    else:
        raise Exception('Warning: Did not save results. A file with the same name already exists!')

# %%
